class.c

```diff
@@ -0,0 +1,1035 @@
+        CvXSUBANY(newcv).any_ptr = stash;
+    }
+
+    /* TODO:
+     *   DOES method
+     */
+
+    struct xpvhv_aux *aux = HvAUX(stash);
+    aux->xhv_class_superclass    = NULL;
+    aux->xhv_class_initfields_cv = NULL;
+    aux->xhv_class_adjust_blocks = NULL;
+    aux->xhv_class_fields        = NULL;
+    aux->xhv_class_next_fieldix  = 0;
+    aux->xhv_class_param_map     = NULL;
+
+    aux->xhv_aux_flags |= HvAUXf_IS_CLASS;
+
+    SAVEDESTRUCTOR_X(invoke_class_seal, stash);
+
+    /* Prepare a suspended compcv for parsing field init expressions */
+    {
+        I32 floor_ix = start_subparse(FALSE, 0);
+
+        CvIsMETHOD_on(PL_compcv);
+
+        /* We don't want to make `$self` visible during the expression but we
+         * still need to give it a name. Make it unusable from pure perl
+         */
+        PADOFFSET padix = pad_add_name_pvs("$(self)", 0, NULL, NULL);
+        assert(padix == PADIX_SELF);
+
+        padix = pad_add_name_pvs("%(params)", 0, NULL, NULL);
+        assert(padix == PADIX_PARAMS);
+
+        PERL_UNUSED_VAR(padix);
+
+        Newx(aux->xhv_class_suspended_initfields_compcv, 1, struct suspended_compcv);
+        suspend_compcv(aux->xhv_class_suspended_initfields_compcv);
+
+        LEAVE_SCOPE(floor_ix);
+    }
+}
+
+#define split_package_ver(value, pkgname, pkgversion)  S_split_package_ver(aTHX_ value, pkgname, pkgversion)
+static const char *S_split_package_ver(pTHX_ SV *value, SV *pkgname, SV *pkgversion)
+{
+    const char *start = SvPVX(value),
+               *p     = start,
+               *end   = start + SvCUR(value);
+
+    while(*p && !isSPACE_utf8_safe(p, end))
+        p += UTF8SKIP(p);
+
+    sv_setpvn(pkgname, start, p - start);
+    if(SvUTF8(value))
+        SvUTF8_on(pkgname);
+
+    while(*p && isSPACE_utf8_safe(p, end))
+        p += UTF8SKIP(p);
+
+    if(*p) {
+        /* scan_version() gets upset about trailing content. We need to extract
+         * exactly what it wants
+         */
+        start = p;
+        if(*p == 'v')
+            p++;
+        while(*p && strchr("0123456789._", *p))
+            p++;
+        SV *tmpsv = newSVpvn(start, p - start);
+        SAVEFREESV(tmpsv);
+
+        scan_version(SvPVX(tmpsv), pkgversion, FALSE);
+    }
+
+    while(*p && isSPACE_utf8_safe(p, end))
+        p += UTF8SKIP(p);
+
+    return p;
+}
+
+#define ensure_module_version(module, version)  S_ensure_module_version(aTHX_ module, version)
+static void S_ensure_module_version(pTHX_ SV *module, SV *version)
+{
+    dSP;
+
+    ENTER;
+
+    PUSHMARK(SP);
+    PUSHs(module);
+    PUSHs(version);
+    PUTBACK;
+
+    call_method("VERSION", G_VOID);
+
+    LEAVE;
+}
+
+#define split_attr_nameval(sv, namp, valp)  S_split_attr_nameval(aTHX_ sv, namp, valp)
+static void S_split_attr_nameval(pTHX_ SV *sv, SV **namp, SV **valp)
+{
+    STRLEN svlen = SvCUR(sv);
+    char *paren_at = (char *)memchr(SvPVX(sv), '(', svlen);
```

Comment:
I believe every `char *` in this function could be `const char *` since you don't modify the SV.

Unfortunately the cast would still be required for C++ builds.

Reply:
They can indeed. Now edited.